pilot/pkg/model/push_context.go

```diff
@@ -512,6 +512,65 @@ func (ps *PushContext) UpdateMetrics() {
 	}
 }
 
+// GatewayServices returns the set of services which are refered from the proxy gateways.
+func (ps *PushContext) GatewayServices(proxy *Proxy) []*Service {
+	svcs := ps.Services(proxy)
+	// gateway set.
+	gateways := map[string]bool{}
+	// host set.
+	hostsFromGateways := map[string]struct{}{}
+
+	for _, gw := range proxy.MergedGateway.GatewayNameForServer {
+		gateways[gw] = true
+	}
+	log.Debugf("GatewayServices: gateway %v has following gw resources:%v", proxy.ID, gateways)
+
+	for _, vsConfig := range ps.VirtualServices(proxy, gateways) {
+		vs, ok := vsConfig.Spec.(*networking.VirtualService)
+		if !ok { // should never happen
+			log.Errorf("Failed in getting a virtual service: %v", vsConfig.Labels)
+			return svcs
```

Comment:
It should never happen, but if it does this is wrong. We should `continue` or return an error and fallback to getting all clusters